lib/demuxeres/gstdemuxeres.c

```diff
@@ -0,0 +1,721 @@
+      return stream;
+    }
+  }
+  return NULL;
+}
+
+static gboolean
+appsink_handle_event (GstDemuxerES * demuxer, GstEvent * event)
+{
+  GstDemuxerESPrivate *priv = demuxer->priv;
+  gboolean ret = FALSE;
+  GST_LOG ("Event %s", GST_EVENT_TYPE_NAME (event));
+  switch (GST_EVENT_TYPE (event)) {
+    case GST_EVENT_STREAM_START:{
+      const gchar *stream_id;
+      GstDemuxerEStream *stream;
+      gst_event_parse_stream_start (event, &stream_id);
+      stream = find_stream (demuxer, stream_id);
+      g_assert (stream);
+      priv->current_stream_id = stream->id;
+      priv->current_stream_type = stream->type;
+      ret = TRUE;
+      break;
+    }
+    case GST_EVENT_STREAM_GROUP_DONE:{
+      set_demuxer_state (demuxer, DEMUXER_ES_STATE_EOS);
+      ret = TRUE;
+      break;
+    }
+    default:
+      GST_LOG ("Unhandled event %s", GST_EVENT_TYPE_NAME (event));
+  }
+  gst_event_unref (event);
+  return ret;
+}
+
+static GstDemuxerESPacket *
+appsink_read_packet (GstDemuxerES * demuxer)
+{
+  GstSample *sample = NULL;
+  GstBuffer *buffer;
+  GstDemuxerESPacket *packet = NULL;
+  GstDemuxerESPrivate *priv = demuxer->priv;
+  GstMiniObject *object;
+
+  if (!priv->pending_sample) {
+    while ((object =
+            gst_app_sink_try_pull_object (GST_APP_SINK (priv->appsink), -1))) {
+      if (GST_IS_EVENT (object)) {
+        appsink_handle_event (demuxer, GST_EVENT (object));
+      } else if (GST_IS_SAMPLE (object)) {
+        sample = GST_SAMPLE (object);
+        break;
+      }
+    }
+  } else {
+    sample = priv->pending_sample;
+    priv->pending_sample = NULL;
+  }
+
+  if (sample) {
+    buffer = gst_sample_get_buffer (sample);
+    if (buffer) {
+      packet = g_new0 (GstDemuxerESPacket, 1);
+      packet->priv = g_new0 (GstDemuxerESPacketPrivate, 1);
+      packet->priv->sample = sample;
+      GstMapInfo mi;
+      gst_buffer_map (buffer, &mi, GST_MAP_READ);
+      packet->data = mi.data;
+      packet->data_size = mi.size;
+      gst_buffer_unmap (buffer, &mi);
+      packet->stream_type = priv->current_stream_type;
+      packet->stream_id = priv->current_stream_id;
+      packet_counter++;
+      packet->packet_number = packet_counter;
+      packet->pts = GST_BUFFER_PTS (buffer);
+      packet->dts = GST_BUFFER_DTS (buffer);
+      packet->duration = GST_BUFFER_DURATION (buffer);
+      GST_LOG ("A new packet for size %ld is available", packet->data_size);
+    }
+  } else {
+    GST_ERROR ("no sample available");
+  }
+  //Look for stream_start or eos event
+  while ((object =
+          gst_app_sink_try_pull_object (GST_APP_SINK (priv->appsink), -1))) {
+    if (GST_IS_EVENT (object)) {
+      if (appsink_handle_event (demuxer, GST_EVENT (object))) {
+        break;
+      }
+    } else if (GST_IS_SAMPLE (object)) {
+      priv->pending_sample = GST_SAMPLE (object);
+      break;
+    }
+  }
+
+  return packet;
+}
+
+static GstDemuxerEStreamType
+gst_parse_stream_get_type_from_pad (GstPad * pad)
```

Comment:
You don't need this if we use GstStream API, which already provides a type: 

https://gstreamer.freedesktop.org/documentation/gstreamer/gststreams.html?gi-language=c#gst_stream_get_stream_type

Reply:
The GstDemuxerEStream object is created on pad added because we need to parse the caps to tell some information about the stream such as  video size etc. That's why it would be complicated to first create the stream on stream start and then update its value on caps event